edx_toggles/toggles/internal/base.py

```diff
@@ -0,0 +1,68 @@
+"""
+Feature toggle base classes
+"""
+
+from abc import ABC
+from weakref import WeakSet
+
+
+class BaseToggle(ABC):
+    """
+    This abstract base class exposes the basic API required by toggle classes. Toggle instances are tracked in the
+    ``_class_instances`` class attribute, which is exposed via the ``get_instaaces`` class method.
+    """
+
+    _class_instances = WeakSet()
+
+    def __init__(self, name, default=False, module_name=""):
+        self.name = name
+        self.default = default
+        self.module_name = module_name
+        self._class_instances.add(self)
+
+    def is_enabled(self):
+        raise NotImplementedError
+
+    @classmethod
+    def get_instances(cls):
+        """
+        Return the list of class instances sorted by name.
+        """
+        return sorted(cls._class_instances, key=lambda instance: instance.name)
+
+
+class BaseNamespace(ABC):
```

Comment:
Please see https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/docs/decisions/0004-waffle-util-namespacing.rst  There are 3 ADRs relevant to the extraction.  I'm sorry if I didn't make that clear.

Reply:
@nasthagiri: FYI: I am planning on the waffle namespace classes going away.  Please see the ADR to understand why.  Since you added namespacing to OEP-17, and were involved in the initial design (I think), it would probably be good for you to read and agree to the [ADR above dropping the namespacing classes](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/docs/decisions/0004-waffle-util-namespacing.rst).

Reply:
@robrap Thanks for the heads up!  I'm more concerned about ensuring developers namespace their toggles, than the mechanism by which we enforce this. Looks like the ADR covers this via an assertion of the existence of a `.` in the name.